### PR TITLE
Simple paging

### PR DESCRIPTION
**Pull request by guy120494**
I have added a new feature for paging.
The relay paging is very nice, but sometime it is overkill.
This new feature is simple paging, meaning you don't have cursors, edges and nodes, only
the arguments (first,last,after,before) and then you directly specify the fields of your entity
The cursors are just numbers, and the fetcher can fetch according to the DB row number.
It is also convenient when the data is sorted according to the DB rows' numbers.

To use it, one needs to annotate his field with `@GraphQLSimpleConnection` and the return type must be an instance of `SimplePaginatedData`. There is a class named `SimplePaginatedDataImpl` to make life easier.

`SimplePaginatedData` has two methods:
getData  - get the data of the relevant entities
getOverAllCount - get the amount of all the entities that can be retrieved

The query for this simple connection would look like that:
```
getHeroesSimpleConnection(first:2, aftet:10){
     overAllCount
     data{
         name
         mana
         hp
      }
}
```

### CODE REVIEW

File: src/main/java/graphql/annotations/connection/AbstractPaginatedData.java [hunk 1]
```diff
@@ -18,9 +18,9 @@
 
 public abstract class AbstractPaginatedData<T> implements PaginatedData<T> {
 
-     boolean hasPreviousPage;
-     boolean hasNextPage;
-     Iterable<T> data;
+    Iterable<T> data;
```
Comment (yarinvak) [inline src/main/java/graphql/annotations/connection/AbstractPaginatedData.java]: should not be private also?

File: src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package graphql.annotations.connection;
+
+public class SimplePaginatedDataImpl<T> extends AbstractSimplePaginatedData<T> {
+
+    private long overAll;
```
Comment (yarinvak) [inline src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java]: whats overall means?

File: src/main/java/graphql/annotations/processor/retrievers/GraphQLFieldRetriever.java [hunk 3]
```diff
@@ -175,6 +189,31 @@ private TypeFunction getTypeFunction(Field field, ProcessingElementsContainer co
         return typeFunction;
     }
 
+    private List<GraphQLArgument> getArgumentsForSimpleConnection() {
+        List<GraphQLArgument> args = new ArrayList<>();
```
Comment (yarinvak) [inline src/main/java/graphql/annotations/processor/retrievers/GraphQLFieldRetriever.java:197]: In simple pagination we want to receive only index & number of elements to receive

File: src/main/java/graphql/annotations/processor/retrievers/fieldBuilders/field/FieldDataFetcherBuilder.java [hunk 4]
```diff
@@ -33,23 +35,25 @@
...
 
+@SuppressWarnings("ALL")
 public class FieldDataFetcherBuilder implements Builder<DataFetcher> {
     private Field field;
     private DataFetcherConstructor dataFetcherConstructor;
     private GraphQLType outputType;
     private TypeFunction typeFunction;
     private ProcessingElementsContainer container;
-    private boolean isConnection;
+    private TypesConnectionChecker typesConnectionChecker;
```
Comment (yarinvak) [inline src/main/java/graphql/annotations/processor/retrievers/fieldBuilders/field/FieldDataFetcherBuilder.java:48]: why not using a "connectionType" field or something like that? and it will be null if no connection is sent..
Because the TypesConnectionChecker is a little bit confusing

**Review by yarinvak — CHANGES_REQUESTED** [review]
The "SimplePagination"/"SimpleConnection" name is a little bit confusing. 
I would suggest calling it @GraphQLChunk
and "SimpleChunk" instead of connection.

**Comment by guy120494** [comment]
On the way I organized some of the classes, that's why There are a lot of modified classes

File: src/main/java/graphql/annotations/connection/AbstractPaginatedData.java (hunk 1 above)
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/AbstractPaginatedData.java]: I don't think so, because the extending class may use the data (not like the pages, where there is no need to use them once you know if they are true or false)

File: src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java (hunk 2 above)
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java]: the amount of all the entities that can be retrieved

File: src/main/java/graphql/annotations/processor/retrievers/GraphQLFieldRetriever.java (hunk 3 above)
Comment (guy120494) [reply src/main/java/graphql/annotations/processor/retrievers/GraphQLFieldRetriever.java:197]: the index is "after" and the number is "first"

**Comment by guy120494** [comment]
@tdraier any thoughts?

File: src/main/java/graphql/annotations/processor/retrievers/GraphQLFieldRetriever.java (hunk 3 above)
Comment (yarinvak) [reply src/main/java/graphql/annotations/processor/retrievers/GraphQLFieldRetriever.java:197]: but there is no need in before and last
and after does not mean "starting index" so its not compatible

File: src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java (hunk 2 above)
Comment (yarinvak) [reply src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java]: call it totalCount or overAllCount

File: src/main/java/graphql/annotations/connection/AbstractPaginatedData.java (hunk 1 above)
Comment (yarinvak) [reply src/main/java/graphql/annotations/connection/AbstractPaginatedData.java]: so set it to protected

Comment (guy120494) [reply src/main/java/graphql/annotations/connection/AbstractPaginatedData.java]: done

File: src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java (hunk 2 above)
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/SimplePaginatedDataImpl.java]: done

**Comment by tdraier** [comment]
I actually have the same kind of needs, but solved it by using a custom relay implementation. The custom relay add some new arguments (instead of first/after, i have limit/offset), the data fetchers still have the @GraphQLConnection with a specific ConnectionFetcher class, that return an extension of Connection with a "nodes" fields (equivalent to your data field). The annotated method returns a object that inherits from PaginatedData (mainly to add the "totalCount", which is not available by default)
The difference is that my connection still have the "default" fields since I inherit from ConnectionFetcher, but you could also remove the "edges" fields and the parameters you don't want by just changing the relay implementation.
Basically it's already possible to customize connections the way you want, without having to change anything in graphql-java-annotation itself. So I'm not sure of the need to add complexity with a new annotation.

**Comment by guy120494** [comment]
That's a good idea, but I don't like the way we change the relay implementation, because what if I want both the default relay for some mutations, and a custom relay for others?
Maybe I should add the relay type in the `@GrahpQLConnection` annotation?

**Comment by tdraier** [comment]
Yes, that's a limitation, currently the same relay implementation is used for all connections. I think it's completely possible to pass the relay implementation to the @GraphQLConnection , and use the one in the container as the default - that would be a nice improvement.

**Comment by guy120494** [comment]
So I'm rejecting this PR, and will open a new one
